### ISSUE TITLE
DaoAuthenticationProvider is not usable on RHEL 8.7 with enforced FIPS mode

### ISSUE THREAD

**Issue by psvo** [spring-projects/spring-security#12873]
**Describe the bug**

Creating instance of DaoAuthenticationProvider fails due to "PBKDF2WithHmacSHA256 SecretKeyFactory not available" when running on RHEL 8.7 with enforced FIPS mode.

The problem is that the DaoAuthenticationProvider creates a default delegating password encoder and one of the delegates fails to instantiate due to limited JCE provider availability when FIPS is enforced.

There's no workaround, because the DaoAuthenticationProvider has only the default constructor which fails due to unconditionally calling `org.springframework.security.crypto.factory.PasswordEncoderFactories#createDelegatingPasswordEncoder`.

The error is:
```
org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.authentication.dao.DaoAuthenticationProvider]: Constructor threw exception; nested exception is java.lang.IllegalArgumentException: Invalid algorithm 'PBKDF2WithHmacSHA256'.
	... more
Caused by: java.security.NoSuchAlgorithmException: PBKDF2WithHmacSHA256 SecretKeyFactory not available
	at javax.crypto.SecretKeyFactory.<init>(SecretKeyFactory.java:122) ~[?:?]
	at javax.crypto.SecretKeyFactory.getInstance(SecretKeyFactory.java:168) ~[?:?]
	at org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.setAlgorithm(Pbkdf2PasswordEncoder.java:226) ~[spring-security-crypto-5.8.2.jar:5.8.2]
	at org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.<init>(Pbkdf2PasswordEncoder.java:179) ~[spring-security-crypto-5.8.2.jar:5.8.2]
	at org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.defaultsForSpringSecurity_v5_8(Pbkdf2PasswordEncoder.java:207) ~[spring-security-crypto-5.8.2.jar:5.8.2]
	at org.springframework.security.crypto.factory.PasswordEncoderFactories.createDelegatingPasswordEncoder(PasswordEncoderFactories.java:81) ~[spring-security-crypto-5.8.2.jar:5.8.2]
	at org.springframework.security.authentication.dao.DaoAuthenticationProvider.<init>(DaoAuthenticationProvider.java:64) ~[spring-security-core-5.8.2.jar:5.8.2]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	... more
```

This is a regression in 5.8.2. It worked for us in Spring Security 5.7.4, because we were overwriting the default password encoder before it tried to retrieve the algorithm.

The issue was most probably introduced by PR #11904 (c50441b59ffa8f34a4f26e341aa109f9b8508372) as a fix for issue #10489.

**To Reproduce**

1. Write an application using DaoAuthenticationProvider from Spring Security 5.8.2.
2. Run the application on RHEL 8.7 with enforced FIPS mode for JDK 11.

**Expected behavior**

Be able to use DaoAuthenticationProvider on RHEL 8.7 with enforced FIPS mode for JDK 11.

Allow providing a custom PasswordEncoder when instantiating DaoAuthenticationProvider to bypass the instantiation of the default delegating password encoder (PasswordEncoderFactories#createDelegatingPasswordEncoder).

**Comment by SeanCodeMedia** [comment]
Hi I am experiencing the same issue when deploying to Open shift. Here is my error 

`Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.authentication.AuthenticationProvider]: Factory method 'daoAuthenticationProvider' threw exception with message: Invalid algorithm 'PBKDF2WithHmacSHA256'.

        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:171) ~[spring-beans-6.0.2.jar!/:6.0.2]

        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:648) ~[spring-beans-6.0.2.jar!/:6.0.2] 

Caused by: java.lang.IllegalArgumentException: Invalid algorithm 'PBKDF2WithHmacSHA256'.

        at org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.setAlgorithm(Pbkdf2PasswordEncoder.java:178) ~[spring-security-crypto-6.0.0.jar!/:6.0.0]

        at org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.<init>(Pbkdf2PasswordEncoder.java:127) ~[spring-security-crypto-6.0.0.jar!/:6.0.0]

        at org.springframework.security.crypto.password.Pbkdf2PasswordEncoder.defaultsForSpringSecurity_v5_8(Pbkdf2PasswordEncoder.java:155) ~[spring-security-crypto-6.0.0.jar!/:6.0.0]

        at org.springframework.security.crypto.factory.PasswordEncoderFactories.createDelegatingPasswordEncoder(PasswordEncoderFactories.java:81) ~[spring-security-crypto-6.0.0.jar!/:6.0.0]

 at org.springframework.security.authentication.dao.DaoAuthenticationProvider.<init>(DaoAuthenticationProvider.java:63) ~[spring-security-core-6.0.0.jar!/:6.0.0]

        at com.lm.autotime.rotarywebtimeclockmiddletier.Config.SecurityConfig.daoAuthenticationProvider(SecurityConfig.java:83) ~[classes!/:na]

        at com.lm.autotime.rotarywebtimeclockmiddletier.Config.SecurityConfig$$SpringCGLIB$$0.CGLIB$daoAuthenticationProvider$4(<generated>) ~[classes!/:na]

        at com.lm.autotime.rotarywebtimeclockmiddletier.Config.SecurityConfig$$SpringCGLIB$$2.invoke(<generated>) ~[classes!/:na]

        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:257) ~[spring-core-6.0.2.jar!/:6.0.2]

        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-6.0.2.jar!/:6.0.2]

        at com.lm.autotime.rotarywebtimeclockmiddletier.Config.SecurityConfig$$SpringCGLIB$$0.daoAuthenticationProvider(<generated>) ~[classes!/:na]

        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]

        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]

        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]

        at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]

        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:139) ~[spring-beans-6.0.2.jar!/:6.0.2]

`

I am running JDK 17. Testesd on both Spring Security  6.0.2 and 6.0.0

**Comment by marcusdacoregio** [comment]
Thanks, @psvo, this is now fixed in `5.8.x` via https://github.com/spring-projects/spring-security/commit/d5603a944d0d13bfac4ae0166027394dc016c168. That commit adds a guard around the new algorithm added and prevents an application from breaking.

Please if you can test the SNAPSHOT it would be great to have feedback if it is working fine before the GA.

The new constructor was added to `6.0.x` via https://github.com/spring-projects/spring-security/pull/12874

**Comment by psvo** [comment]
> Please if you can test the SNAPSHOT it would be great to have feedback if it is working fine before the GA.

@marcusdacoregio, I can confirm the fix merged to 5.8.3-SNAPSHOT works for me. Thank you!
